repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeRoutingServlet.scala

```diff
@@ -0,0 +1,362 @@
+      val endpoint = getEndpoint(destination, locationUrl)
+      val servletResponseStatus = servletResponse.getStatus
+      metricsService.get().foreach { metSer =>
+        markResponseCodeHelper(metSer, servletResponseStatus, endpoint)
+        markResponseCodeHelper(metSer, servletResponseStatus, AllEndpoints)
+        markRequestTimeoutHelper(metSer, servletResponseStatus, endpoint)
+        markRequestTimeoutHelper(metSer, servletResponseStatus, AllEndpoints)
+      }
+      val stopTime = System.currentTimeMillis
+      reportingService.recordServiceResponse(routeDest.getDestinationId, servletResponseStatus, stopTime - startTime)
+      responseHeaderService.fixLocationHeader(
+        servletRequest.getRequest.asInstanceOf[HttpServletRequest],
+        servletResponse,
+        routeDest,
+        locationUrl.getPath,
+        destination.getRootPath)
+    } recover {
+      case e: IOException if e.getCause.isInstanceOf[ReadLimitReachedException] =>
+        logger.error("Error reading request content", e)
+        servletResponse.sendError(HttpServletResponse.SC_REQUEST_ENTITY_TOO_LARGE, "Error reading request content")
+      case e =>
+        logger.error("Error communicating with {}", locationUrl.getPath, e)
+        servletResponse.setStatus(HttpServletResponse.SC_SERVICE_UNAVAILABLE)
+    }
+  }
+
+  def getEndpoint(dest: Destination, locationUrl: URL): String = {
+    val sb = new StringBuilder
+    sb.append(locationUrl.getHost).append(":").append(locationUrl.getPort)
+    dest match {
+      case endpoint: DestinationEndpoint => sb.append(endpoint.getRootPath)
+      case cluster: DestinationCluster => sb.append(cluster.getRootPath)
+      case _ => throw new IllegalArgumentException("Unknown destination type: " + dest.getClass.getName)
+    }
+    sb.toString
+  }
+
+  def markResponseCodeHelper(metricsService: MetricsService, responseCode: Int, component: String): Unit = {
+    if (100 < responseCode && responseCode < 600) {
+      metricsService
+        .getRegistry
+        .meter(MetricRegistry.name("org.openrepose.core.ResponseCode", component, "%dXX".format(responseCode / 100)))
+        .mark()
+    } else {
+      logger.error(s"$component Encountered invalid response code: $responseCode")
+    }
+  }
+
+  def markRequestTimeoutHelper(metricsService: MetricsService, responseCode: Int, endpoint: String): Unit = {
+    if (responseCode == SC_REQUEST_TIMEOUT) {
+      metricsService
+        .getRegistry
+        .meter(MetricRegistry.name("org.openrepose.core.RequestTimeout.TimeoutToOrigin", endpoint))
+        .mark()
+    }
+  }
+}
+
+object ReposeRoutingServlet {
+  final val SystemModelConfigurationFilename = "system-model.cfg.xml"
+  final val NotInitializedMessage = "The ReposeRoutingServlet has not yet been initialized"
+  final val FailedToRouteMessage = "Failed to route the request to the origin service"
+  private val AllEndpoints = "All Endpoints"
+  private val HttpsProtocol = "https"
```

Comment:
If a variable is only marked as `val`, then the sub-classes can `override` them.
So the use of the `final` prefix prevents that.
By definition a `private val` is already a `final val` since a `private val` in a companion object can not be seen by sub-classes of it to be `override`'n.

Reply:
Ok.